### PR TITLE
Add dumping commands to jcmd

### PR DESCRIPTION
**Pull request by pdbain-ibm**
Support various forms of dumps via jcmd.  Include the legacy format for heap dumps.
Clean up some whitespace and other tidying.

Add unit tests.

This is part of https://github.com/eclipse/openj9/issues/5164

Signed-off-by: Peter Bain <peter_bain@ca.ibm.com>

### CODE REVIEW

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticProperties.java [hunk 1]
```diff
@@ -374,9 +378,10 @@ public static Properties makeExceptionProperties(Exception e) {
 		props.put(IPC.PROPERTY_DIAGNOSTICS_ERROR, Boolean.toString(true));
 		props.put(IPC.PROPERTY_DIAGNOSTICS_ERRORTYPE, e.getClass().getName());
 		String msg = e.getMessage();
-		if (null != msg) {
-			props.put(IPC.PROPERTY_DIAGNOSTICS_ERRORMSG, msg);
+		if (null == msg) {
+			msg = "Error at " + e.getStackTrace()[0].toString(); //$NON-NLS-1$
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticProperties.java]: I think `toString()` here is not required. Also I would suggest `String.format()` over String concatenation though here is probably a slow path anyway.

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java [hunk 2]
```diff
@@ -68,6 +80,15 @@
...
 	public static final String DIAGNOSTICS_GC_CLASS_HISTOGRAM = "GC.class_histogram"; //$NON-NLS-1$
 
+	/**
+	 * Commands to generate dumps of various types
+	 */
+	public static final String DIAGNOSTICS_DUMP_HEAP = "Dump.heap"; //$NON-NLS-1$
+	public static final String DIAGNOSTICS_GC_HEAP_DUMP = "GC.heap_dump"; //$NON-NLS-1$
+	public static final String DIAGNOSTICS_DUMP_JAVA = "Dump.java"; //$NON-NLS-1$
+	public static final String DIAGNOSTICS_DUMP_SNAP = "Dump.snap"; //$NON-NLS-1$
+	public static final String DIAGNOSTICS_DUMP_SYSTEM = "Dump.system"; //$NON-NLS-1$
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: Is there a plan to refer these declaration outside? otherwise they can be (package) private.

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java [hunk 3]
```diff
@@ -207,7 +228,36 @@ private static DiagnosticProperties getThreadInfo(String diagnosticCommand) {
...
+	}
+
+	private static DiagnosticProperties doDump(String diagnosticCommand) {
+		DiagnosticProperties result = null;
+		String[] parts = diagnosticCommand.split(DIAGNOSTICS_OPTION_SEPARATOR);
+		logMessage("doDump: ", diagnosticCommand); //$NON-NLS-1$
+		if (parts.length != 2) {
+			result =  makeErrorProperties("Error: wrong number of arguments"); //$NON-NLS-1$
+		} else {
+			String pathString = parts[1];
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: This variable is referred only once, `parts[1]` can be used directly instead.

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java [hunk 4]
```diff
@@ -207,7 +228,36 @@ private static DiagnosticProperties getThreadInfo(String diagnosticCommand) {
...
+			String dumpType;
+			/* handle legacy form of Dump.heap for compatibility with reference implementation */
+			if (DIAGNOSTICS_GC_HEAP_DUMP.equals(parts[0])) {
+				dumpType = "heap"; //$NON-NLS-1$
+			} else {
+				dumpType = parts[0].split("\\.")[1]; //$NON-NLS-1$
+			}
+			String fileDirective = ("system".equals(dumpType) &&  IPC.isZOS) ? ":dsn=" : ":file="; //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: nitpicking: there is an extra space before `IPC.isZOS`.

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java [hunk 5]
```diff
@@ -207,7 +228,36 @@ private static DiagnosticProperties getThreadInfo(String diagnosticCommand) {
...
+
+			String request = dumpType + fileDirective + pathString;
+			try {
+				String actualDumpFile = triggerDumpsImpl(request, dumpType + "DumpToFile"); //$NON-NLS-1$
+				result = makeStringResult("Dump written to " + actualDumpFile); //$NON-NLS-1$
+			} catch (InvalidDumpOptionExceptionBase e) {
+				logMessage("doDump exception: ", e.getMessage()); //$NON-NLS-1$
+				DiagnosticProperties.makeExceptionProperties(e);
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: Any particularly reason `DiagnosticProperties.makeExceptionProperties()` isn't imported statically like `DiagnosticProperties.makeErrorProperties()` or `DiagnosticProperties.makeStringResult`?

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java [hunk 6]
```diff
@@ -237,49 +287,76 @@ private static DiagnosticProperties doHelp(String diagnosticCommand) {
...
+
+		commandTable.put(DIAGNOSTICS_DUMP_JAVA, DiagnosticUtils::doDump);
+		helpTable.put(DIAGNOSTICS_DUMP_JAVA, DIAGNOSTICS_DUMP_JAVA_HELP);
+
+		commandTable.put(DIAGNOSTICS_DUMP_SNAP, DiagnosticUtils::doDump);
+		helpTable.put(DIAGNOSTICS_DUMP_SNAP, DIAGNOSTICS_DUMP_SNAP_HELP);
+
+		commandTable.put(DIAGNOSTICS_DUMP_SYSTEM, DiagnosticUtils::doDump);
+		helpTable.put(DIAGNOSTICS_DUMP_SYSTEM, DIAGNOSTICS_DUMP_SYSTEM_HELP);
+	}
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: These two `HashMap` can be lazy initialized to help startup performance.

**Comment by keithc-ca** [comment]
Please remove `test/functional/TestUtilities/src/org/openj9/test/util/.StringUtilities.java.swp` from this change.

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java (hunk 5 above)
Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: My thought was about the consistency of `DiagnosticProperties` usages, but would like to listen why `DiagnosticProperties.makeExceptionProperties()` is different from those two.

File: jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java [hunk 7]
```diff
@@ -81,6 +81,11 @@
 	 */
 	public static boolean isWindows = false;
 
+	/**
+	 * True if operating system is z/OS.
+	 */
+	public static boolean isZOS = false;
```
[the remark this answers is not in the document]
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java]: Good plan.  Please open an issue.

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticProperties.java (hunk 1 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticProperties.java]: There is an implicit `toString()`, but I wanted to make it explicit.
I like the String.format() idea.  Done.

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java (hunk 5 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: > Any particularly reason `DiagnosticProperties.makeExceptionProperties()` isn't imported statically like `DiagnosticProperties.makeErrorProperties()` or `DiagnosticProperties.makeStringResult`?

I generally `import static` only items which are used multiple times.

File: jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java]: Opened https://github.com/eclipse/openj9/issues/6710

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java (hunk 5 above)
Comment (keithc-ca) [reply jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: I suggest `import static` should only be used when it doesn't reduce the readability of the code. IMO, none of the uses in the pull request qualify. An unqualified reference is more work for the (human) reader to understand.

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java (hunk 6 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: This class is not loaded unless there is a diagnostic operation, so startup will not be affected.

**Comment by pdbain-ibm** [comment]
Updated per reviewers' comments.

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java (hunk 5 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: Thanks for the suggestion, Keith.  I will keep that in mind.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java [hunk 8]
```diff
@@ -51,12 +53,19 @@
 	private static final String GC_RUN = "GC.run"; //$NON-NLS-1$
 	private static final String GC_CLASS_HISTOGRAM = "GC.class_histogram"; //$NON-NLS-1$
 	private static final String JCMD_COMMAND = "jcmd"; //$NON-NLS-1$
-	private static String[] JCMD_COMMANDS = { GC_CLASS_HISTOGRAM, GC_RUN, THREAD_PRINT, HELP_COMMAND };
+	public static final String DUMP_HEAP = "Dump.heap"; //$NON-NLS-1$
+	public static final String GC_HEAP_DUMP = "GC.heap_dump"; //$NON-NLS-1$
+	public static final String DUMP_JAVA = "Dump.java"; //$NON-NLS-1$
+	public static final String DUMP_SNAP = "Dump.snap"; //$NON-NLS-1$
+	public static final String DUMP_SYSTEM = "Dump.system"; //$NON-NLS-1$
```
Comment (JasonFengJ9) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: These `public` variables can be `private`.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java [hunk 9]
```diff
@@ -51,12 +53,19 @@
 	private static final String GC_RUN = "GC.run"; //$NON-NLS-1$
 	private static final String GC_CLASS_HISTOGRAM = "GC.class_histogram"; //$NON-NLS-1$
 	private static final String JCMD_COMMAND = "jcmd"; //$NON-NLS-1$
-	private static String[] JCMD_COMMANDS = { GC_CLASS_HISTOGRAM, GC_RUN, THREAD_PRINT, HELP_COMMAND };
+	public static final String DUMP_HEAP = "Dump.heap"; //$NON-NLS-1$
+	public static final String GC_HEAP_DUMP = "GC.heap_dump"; //$NON-NLS-1$
```
Comment (JasonFengJ9) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: This define can be moved after `DUMP_SYSTEM` to match the order within `JCMD_COMMANDS`; similarly, `GC_CLASS_HISTOGRAM` should be before `GC_RUN`. Another option is to re-arrange the defines in alphabetical order, and update `JCMD_COMMANDS` in same order as well.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java [hunk 10]
```diff
@@ -122,7 +131,7 @@ public void testCommandNoOptions() throws IOException {
 			String expectedString = commandExpectedOutputs.getOrDefault(command, "Test error: expected output not defined"); //$NON-NLS-1$
 			log("Expected string: " + expectedString); //$NON-NLS-1$
 			Optional<String> searchResult = StringUtilities.searchSubstring(expectedString, jcmdOutput);
-			assertTrue(searchResult.isPresent(), "Expected string not found " + expectedString); //$NON-NLS-1$
+			assertTrue(searchResult.isPresent(), "Expected string \""  + expectedString + "\" not found "); //$NON-NLS-1$ //$NON-NLS-2$
```
Comment (JasonFengJ9) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: nitpicking: extra space after _not found_.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java [hunk 11]
```diff
@@ -163,6 +172,48 @@ public void testClassHistogramAll() throws IOException {
...
+				assertTrue(searchResult.isPresent(), "Expected string not found in jcmd output: " + expectedString); //$NON-NLS-1$
+				log(EXPECTED_STRING_FOUND + " in jcmd output: " + expectedString); //$NON-NLS-1$
+
+				expectedString = dumpType + " dump written to " + dumpFilePath; //$NON-NLS-1$
+				searchResult = StringUtilities.searchSubstring(expectedString,tgt.getTargetErrReader().lines());
+				assertTrue(searchResult.isPresent(), "Expected string not found in target standard error: " + expectedString); //$NON-NLS-1$
+				log(EXPECTED_STRING_FOUND + " in target standard error: " + expectedString); //$NON-NLS-1$
+			} finally {
+				tgt.terminateTarget();
+				dumpFileLocation.delete();
```
Comment (JasonFengJ9) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: Is this redundant since same call is made right after `new File(userDir, "my" + dumpType + "Dump")`?

Comment (JasonFengJ9) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: Additional question, is it required to delete the folder `dumpFileLocation` for each sub-test? Can all sub-tests share same dump file folder? Can the `TargetManager tgt` be re-used?

Comment (pdbain-ibm) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: > Is this redundant since same call is made right after `new File(userDir, "my" + dumpType + "Dump")`?

If a previous run did not clean up, e.g. due to a crash, I still want to make sure the directory is clean, since the test will fail if the requested file already exists.


> Additional question, is it required to delete the folder `dumpFileLocation` for each sub-test? 

Yes: the test will fail if the file already exists, since in that case the dump agent chooses another file in another directory. 

> Can all sub-tests share same dump file folder? 

They already do.

> Can the `TargetManager tgt` be re-used?

Possibly, though it makes the code more complicated (extra `try/finally`).  Launching and terminating the target is quick.

**Comment by pdbain-ibm** [comment]
Updated per @JasonFengJ9 's comments.

File: jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java [hunk 12]
```diff
@@ -149,6 +149,7 @@ static void initializeAttachAPI() {
 		String osName = com.ibm.oti.vm.VM.getVMLangAccess().internalGetProperties().getProperty("os.name"); //$NON-NLS-1$
 		if (null != osName) {
 			if (osName.equalsIgnoreCase("z/OS")) { //$NON-NLS-1$
+				IPC.isZOS = true;
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java]: This doesn't belong here: `IPC` should define the values of its own fields.

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java [hunk 13]
```diff
@@ -23,6 +23,10 @@
 
 package openj9.tools.attach.diagnostics.base;
 
+import static com.ibm.tools.attach.target.IPC.logMessage;
+import static openj9.tools.attach.diagnostics.base.DiagnosticProperties.makeErrorProperties;
+import static openj9.tools.attach.diagnostics.base.DiagnosticProperties.makeStringResult;
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: Please don't use static imports like this: it only makes the code harder to understand.

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java [hunk 14]
```diff
@@ -207,7 +228,35 @@ private static DiagnosticProperties getThreadInfo(String diagnosticCommand) {
...
+		logMessage("doDump: ", diagnosticCommand); //$NON-NLS-1$
+		if (parts.length != 2) {
+			result =  makeErrorProperties("Error: wrong number of arguments"); //$NON-NLS-1$
+		} else {
+			String dumpType;
+			/* handle legacy form of Dump.heap for compatibility with reference implementation */
+			if (DIAGNOSTICS_GC_HEAP_DUMP.equals(parts[0])) {
+				dumpType = "heap"; //$NON-NLS-1$
+			} else {
+				dumpType = parts[0].split("\\.")[1]; //$NON-NLS-1$
```
Comment (keithc-ca) [inline jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: How can this be sure that `parts[0]` contains any `'.'` characters?

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java (hunk 10 above)
Comment (keithc-ca) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: more nits: extra spaces before `+`

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java [hunk 15]
```diff
@@ -163,6 +173,48 @@ public void testClassHistogramAll() throws IOException {
 		log(EXPECTED_STRING_FOUND);
 	}
 
+	@Test
+	public void testDumps() throws IOException {
+		@SuppressWarnings("nls")
+		String[][] commandsAndDumpTypes = {{DUMP_HEAP, "Heap"}, {GC_HEAP_DUMP, "Heap"}, {DUMP_JAVA, "Java"}, {DUMP_SNAP, "Snap"}, {DUMP_SYSTEM, "System"}};
+		for (String[] commandAndDumpName: commandsAndDumpTypes) {
```
Comment (keithc-ca) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: space both sides of `:`

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java [hunk 16]
```diff
@@ -163,6 +173,48 @@ public void testClassHistogramAll() throws IOException {
...
+			TargetManager tgt = new TargetManager(TestConstants.TARGET_VM_CLASS, null,
+					Collections.emptyList(), Collections.singletonList("-Xmx10M")); //$NON-NLS-1$
+			tgt.syncWithTarget();
+			String targetId = tgt.targetId;
+			assertNotNull(targetId, "target did not launch"); //$NON-NLS-1$
+			String dumpType = commandAndDumpName[1];
+			@SuppressWarnings("nls")
```
Comment (keithc-ca) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: This annotation might be better applied to the method (or the class); mixing both annotations and `//$NON-NLS-` comments seems wrong (unless they're a small minority of cases).

File: test/functional/TestUtilities/src/org/openj9/test/util/StringUtilities.java [hunk 17]
```diff
@@ -38,10 +39,19 @@
...
+		return searchSubstring(needle, haystack.stream());
+	}
+	
+	/**
+	 * Find the first occurrence of a substring in a list of Strings.
+	 * @param needle pattern for which to search
+	 * @param haystack list of Strings to search
+	 * @return Optional which is either empty or contains matching string 
+	 */
+	public static Optional<String> searchSubstring(String needle, Stream<String> haystack) {
```
Comment (keithc-ca) [inline test/functional/TestUtilities/src/org/openj9/test/util/StringUtilities.java]: Is this new method actually used (other than immediately above)? If you generalize the first method so the haystack has type `Collection<String>`, it becomes more useful, perhaps to the point where this method is no longer needed.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java [hunk 18]
```diff
@@ -172,12 +224,20 @@ protected void setUp(Method testMethod) {
 	@BeforeSuite
 	protected void setupSuite() {
 		assertTrue(PlatformInfo.isOpenJ9(), "This test is valid only on OpenJ9"); //$NON-NLS-1$
+		userDir = new File(System.getProperty("user.dir")); //$NON-NLS-1$
 		getJdkUtilityPath(JCMD_COMMAND);
 		commandExpectedOutputs = new HashMap<>();
 		commandExpectedOutputs.put(HELP_COMMAND, THREAD_PRINT);
```
Comment (JasonFengJ9) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: This is not part of this PR however it seems wrong.

File: test/functional/TestUtilities/src/org/openj9/test/util/StringUtilities.java [hunk 19]
```diff
@@ -38,10 +39,19 @@
...
+		return haystack.filter(s -> s.contains(needle)).findFirst();
```
Comment (JasonFengJ9) [inline test/functional/TestUtilities/src/org/openj9/test/util/StringUtilities.java]: Since all references to this method just need `searchResult.isPresent()`, this method can return a boolean result instead.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java [hunk 20]
```diff
@@ -122,7 +132,7 @@ public void testCommandNoOptions() throws IOException {
 			String expectedString = commandExpectedOutputs.getOrDefault(command, "Test error: expected output not defined"); //$NON-NLS-1$
```
Comment (JasonFengJ9) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: Should the test fail here immediately if the expected output wasn't defined?

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java (hunk 14 above)
Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: True.  Needs some error checking...

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java (hunk 16 above)
Comment (pdbain-ibm) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: I believe the preferred practice is to add the metacomments individually where practical.I used the annotation where there is a large number of literals on a line to avoid a large number of metacomments.

File: test/functional/TestUtilities/src/org/openj9/test/util/StringUtilities.java (hunk 17 above)
Comment (pdbain-ibm) [reply test/functional/TestUtilities/src/org/openj9/test/util/StringUtilities.java]: > Is this new method actually used (other than immediately above)? 

Yes, it is used in `testDumps()` to scan a stream of input lines.

> If you generalize the first method so the haystack has type `Collection<String>`, it becomes more useful

Good idea.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java (hunk 18 above)
Comment (pdbain-ibm) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: This is intended: the test needs to know the current working directory.

File: test/functional/TestUtilities/src/org/openj9/test/util/StringUtilities.java (hunk 19 above)
Comment (pdbain-ibm) [reply test/functional/TestUtilities/src/org/openj9/test/util/StringUtilities.java]: This is intended as a general utility.  Callers may want the full String containing the needle.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java (hunk 20 above)
Comment (pdbain-ibm) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: It is a test coding error if the expected output is not defined, and the test will fail because the expected output is bogus.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java (hunk 16 above)
Comment (keithc-ca) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: I don't think we have any NLS concerns in test code: don't bother with the metacomments or add an annotation to the class.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java (hunk 18 above)
Comment (keithc-ca) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: `new File(".")` is a much simpler way to refer to the current directory.

**Comment by pdbain-ibm** [comment]
Updated and tested manually.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java (hunk 18 above)
Comment (JasonFengJ9) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: Not sure about the response. My question was about `commandExpectedOutputs.put(HELP_COMMAND, THREAD_PRINT);`, why is `THREAD_PRINT` the expected outputs of `HELP_COMMAND`?

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java (hunk 16 above)
Comment (pdbain-ibm) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: @DanHeidinga @pshipton Do you concur with @keithc-ca's request?

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java (hunk 18 above)
Comment (pdbain-ibm) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: > why is `THREAD_PRINT` the expected outputs of `HELP_COMMAND`?

`THREAD_PRINT` is _part_ of the expected output of `HELP_COMMAND`.  I chose it
as a representative sample of the output.

File: jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java [hunk 21]
```diff
@@ -145,15 +145,7 @@ static void initializeAttachAPI() {
...
-		boolean enableAttach = true;
-		String osName = com.ibm.oti.vm.VM.getVMLangAccess().internalGetProperties().getProperty("os.name"); //$NON-NLS-1$
-		if (null != osName) {
-			if (osName.equalsIgnoreCase("z/OS")) { //$NON-NLS-1$
-				enableAttach = false;
-			} else if (osName.startsWith("Windows")) { //$NON-NLS-1$
-				IPC.isWindows = true;
-			}
-		}
+		boolean enableAttach = ! IPC.isZOS;
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/com/ibm/tools/attach/target/AttachHandler.java]: nitpicking: extra space between `!` and `IPC.isZOS`.

File: jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java [hunk 22]
```diff
@@ -81,6 +81,22 @@
...
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java]: Can `isZOS` be `final` or accessed via `getter` instead? Similarly question for `isWindows`.

File: jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java [hunk 23]
```diff
@@ -81,6 +81,22 @@
...
+
+	static {
+		String osName = com.ibm.oti.vm.VM.getVMLangAccess().internalGetProperties().getProperty("os.name"); //$NON-NLS-1$
+		if (null != osName) {
+			if (osName.equalsIgnoreCase("z/OS")) { //$NON-NLS-1$
+				IPC.isZOS = true;
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java]: There is no need to specify current class `IPC`, similar for `IPC.isWindows` below.

File: jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java (hunk 22 above)
Comment (keithc-ca) [reply jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java]: Please arrange for the fields to be `final`.

Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/com/ibm/tools/attach/target/IPC.java]: > Can `isZOS` be `final`? 

Done

File: jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java [hunk 24]
```diff
@@ -212,14 +229,49 @@ private static DiagnosticProperties getThreadInfo(String diagnosticCommand) {
 		return result;
 	}
 
+	private static DiagnosticProperties doDump(String diagnosticCommand) {
+		DiagnosticProperties result = null;
+		String[] parts = diagnosticCommand.split(DIAGNOSTICS_OPTION_SEPARATOR);
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: Is there a test against a `diagnosticCommand` with `DIAGNOSTICS_OPTION_SEPARATOR`?

Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: Yes,  for example `testThreadPrint()`.

Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: Since `DIAGNOSTICS_OPTION_SEPARATOR = ","`, I assume there is a `,` within the test command options however  It doesn't seem obvious within `testThreadPrint()`.
In addition, `testThreadPrint()` is not part of this PR, does the newly added test provide code coverage for the code path when `DIAGNOSTICS_OPTION_SEPARATOR` presents?

Comment (pdbain-ibm) [reply jcl/src/java.base/share/classes/openj9/tools/attach/diagnostics/base/DiagnosticUtils.java]: The separator is inserted by `DiagnosticUtils.makeJcmdCommand()`. This is exercised by other tests, as well as this test: the separator is used to append the file name.  Running `jcmd <target> Dump.java myJavaDump` with logging enabled, the target receives `doCommand ATTACH_DIAGNOSTICS:Dump.java,myJavaDump`.

**Comment by pdbain-ibm** [comment]
@JasonFengJ9 and @keithc-ca have I answered all your concerns?
Thanks

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java [hunk 25]
```diff
@@ -122,7 +133,7 @@ public void testCommandNoOptions() throws IOException {
...
 			log("Expected string: " + expectedString); //$NON-NLS-1$
 			Optional<String> searchResult = StringUtilities.searchSubstring(expectedString, jcmdOutput);
-			assertTrue(searchResult.isPresent(), "Expected string not found " + expectedString); //$NON-NLS-1$
+			assertTrue(searchResult.isPresent(), "Expected string \"" + expectedString + "\" not found "); //$NON-NLS-1$ //$NON-NLS-2$
```
Comment (keithc-ca) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: nit: extra space after 'not found'

Comment (pdbain-ibm) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: Fixed.  Thanks.

**Comment by keithc-ca** [comment]
Jenkins test extended plinux,win jdk11

**Comment by pdbain-ibm** [comment]
Windows failed:
`18:24:06  jvm.def : error LNK2001: unresolved external symbol JVM_InitializeFromArchive`.
Looks like the branch needs rebasing.

**Comment by keithc-ca** [comment]
Jenkins test extended win jdk11

**Comment by keithc-ca** [comment]
Trying Windows build again; I checked that #6751 is included in the merge commit first.

**Comment by keithc-ca** [comment]
@pdbain-ibm Why did you push a new commit? It seems only to be a rebase. The only effect appears to be loss of information on the PR builds.

**Comment by pdbain-ibm** [comment]
> @pdbain-ibm Why did you push a new commit? It seems only to be a rebase. The only effect appears to be loss of information on the PR builds.

I rebased to fix the compilation failure mentioned above; the last rebase was a week ago.

**Comment by keithc-ca** [comment]
Linux ppc64le testing passed earlier; just doing Windows testing now.

**Comment by keithc-ca** [comment]
Jenkins test extended win jdk11

**Comment by keithc-ca** [comment]
Jenkins copyright check

**Comment by keithc-ca** [comment]
Jenkins line endings check

**Comment by keithc-ca** [comment]
Jenkins test extended win jdk11
